client/dfclient/src/loading_state.cpp

```diff
@@ -0,0 +1,164 @@
+	_outline->setPosition(res.x*0.5f, res.y*0.6f);
+	_outline->setLayer(3);
+}
+
+StateType LoadingState::Update(Messages m) {
+	// load a single file
+	std::string path = "";
+	if (!_textureFilenames.empty()){
+		path = _textureFilenames.back();
+		std::string filename = ncine::FileSystem::baseName(path.c_str()).data();
+		_resources.textures[filename] = std::make_unique<ncine::Texture>(path.c_str());
+		_textureFilenames.pop_back();
+
+	} else if (!_soundFilenames.empty()) {
+		path = _soundFilenames.back();
+		std::string filename = ncine::FileSystem::baseName(path.c_str()).data();
+		_resources._sounds[filename] = std::make_unique<ncine::AudioBuffer>(path.c_str());
+		_soundFilenames.pop_back();
+	}
+	_current += fileSize(path);
+
+	// update the loading text
+	auto _percent_loaded = _current/(double)_total * 100;
+	auto string = "Loading... " + std::to_string((int)_percent_loaded) + "%";
+	_loadingText->setString(string.c_str());
+	unsigned char alpha = 55 + 2*_percent_loaded;
+	_loadingText->setAlpha(alpha);
+
+	// update the loading graphics
+	_curtain->setPosition(0, 1.2*_percent_loaded);
+
+	// manage the ending
+	if (_percent_loaded == 100) {
+		if (_fadeoutTween == nullptr) {
+			// stretch curtain to cover the whole screen
+			ncine::Vector2f res = {ncine::theApplication().width(), ncine::theApplication().height()};
+			std::vector<ncine::Vector2f> vertices(4);	
+			vertices[0].x = 0;
+			vertices[0].y = -1.2*_percent_loaded;
+			vertices[1].x = 0;
+			vertices[1].y = res.y + 1 - 1.2*_percent_loaded;
+			vertices[2].x = res.x + 1;
+			vertices[2].y = -1.2*_percent_loaded;
+			vertices[3].x = res.x + 1;
+			vertices[3].y = res.y + 1 - 1.2*_percent_loaded;
+			_curtain->createVerticesFromTexels(vertices.size(), vertices.data());
+			_curtain->setPosition(0, 0);
+			_curtain->setLayer(65535);
+			_curtain->setAlpha(0);
+
+			// create fadeout tween
+			ncine::MeshSprite* curtainPtr = _curtain.get();
+			auto fadeoutTween = tweeny::from(0).to(255).during(25).via(tweeny::easing::circularOut)
+				.onStep([curtainPtr] (tweeny::tween<int>& t, int v) -> bool {
+					curtainPtr->setAlpha(v);
+					return false;
+				});
+			_resources._intTweens.push_back(std::move(fadeoutTween));
+			_fadeoutTween = &(*(_resources._intTweens.rbegin()));  // ... -_-
+
+		} else if (_fadeoutTween->progress() == 1) {
+			return StateType::Menu;
+		} 
+	}
+	return StateType::Loading;
+}
+
+LoadingState::~LoadingState() {
+	// init soundbuffers
+	_resources._killSoundBuffer = std::move(_resources._sounds["amongus-kill.wav"]);
+	_resources._sounds.erase("amongus-kill.wav");
+	_resources._killSound = std::make_unique<ncine::AudioBufferPlayer>(_resources._killSoundBuffer.get());
+
+	_resources._goldfishSoundBuffer = std::move(_resources._sounds["mario-powerup.wav"]);
+	_resources._sounds.erase("mario-powerup.wav");
+	_resources._goldfishSound = std::make_unique<ncine::AudioBufferPlayer>(_resources._goldfishSoundBuffer.get());
```

Comment:
why is there some initialization logic in the _destructor_

Reply:
Because this way we know it will be executed and that it will be executed lastly. It cannot be in the Update method (_maybe_ it could be placed after an if-gate that makes sure everything is finished loading, but I don't like the idea of checking if we are done after every file) and the destructor of the LoadingState seems like a good transition place.

Reply:
>I don't like the idea of checking if we are done after every file

we do it anyways
```
if (_percent_loaded == 100) {
```
I'm not going to make you commit this trivial change yourself, but I will change it